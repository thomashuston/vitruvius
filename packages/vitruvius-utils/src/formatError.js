import * as path from 'path';
import chalk from 'chalk';

const BABEL_MESSAGE_PATTERN = /^([^:]+): (.*) \([0-9]+:[0-9]+\)$/;
const ROOT_DIR = process.cwd();

function separateMessageFromFile(message) {
    const messageMatch = message.match(BABEL_MESSAGE_PATTERN);
    if (messageMatch) {
        return messageMatch[2];
    } else {
        return message;
    }
}

export default function formatError(error, file) {
    if (error._babel) {
        const message = separateMessageFromFile(error.message);
        const relativeFilePath = path.relative(ROOT_DIR, file);
        return `${message}\n\n at ${chalk.reset.cyan(relativeFilePath)}:${error.loc.line}:${error.loc.column}\n\n${error.codeFrame}`;
    } else {
        return error.toString();
    }
}
