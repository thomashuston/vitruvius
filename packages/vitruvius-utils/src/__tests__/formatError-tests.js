import * as path from 'path';
import { transform } from 'babel-core';
import formatError from '../formatError';

const getError = (fn) => {
    try {
        fn();
    } catch (e) {
        return e;
    }
};

it('formats Babel errors', () => {
    const error = getError(() => transform('cons invalid = true;'));

    expect(
        formatError(error, path.join(process.cwd(), 'foo/bar.js'))
    ).toMatchSnapshot();
});

it('returns other errors as strings', () => {
    const error = getError(() => { throw new Error('other error type'); });

    expect(
        formatError(error, path.join(process.cwd(), 'foo/bar.js'))
    ).toMatchSnapshot();
});
