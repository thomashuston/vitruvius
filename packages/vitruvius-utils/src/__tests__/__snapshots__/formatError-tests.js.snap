// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formats Babel errors 1`] = `
"Unexpected token, expected ;

 at [0m[36mfoo/bar.js[39m[0m:1:5

[0m[31m[1m>[22m[39m[90m 1 | [39mcons invalid [33m=[39m [36mtrue[39m[33m;[39m
 [90m   | [39m     [31m[1m^[22m[39m[0m"
`;

exports[`returns other errors as strings 1`] = `"Error: other error type"`;
